src/docfx/build/metadata/ContributionInfo.cs

```diff
@@ -36,17 +36,32 @@ public static ContributionInfo Load(Docset docset)
         {
             var commitsByFile = LoadCommits(docset);
 
-            var updateTimeByCommit = string.IsNullOrEmpty(docset.Config.Contribution.GitCommitsTimePath)
+            var updateTimeByCommit = string.IsNullOrEmpty(docset.Config.Contribution.GitCommitsTimeUrl)
                 ? new Dictionary<string, DateTime>()
-                : GitCommitsTime.Create(Path.Combine(docset.DocsetPath, docset.Config.Contribution.GitCommitsTimePath)).ToDictionary();
+                : GitCommitsTime.Create(GetRestorePathFromUrl(docset.Config.Contribution.GitCommitsTimeUrl, docset.RestoreMap)).ToDictionary();
 
-            var userProfileCachePath = string.IsNullOrEmpty(docset.Config.Contribution.UserProfileCachePath)
+            var userProfileCachePath = string.IsNullOrEmpty(docset.Config.Contribution.UserProfileCacheUrl)
                 ? s_defaultProfilePath
-                : Path.Combine(docset.DocsetPath, docset.Config.Contribution.UserProfileCachePath);
+                : GetRestorePathFromUrl(docset.Config.Contribution.UserProfileCacheUrl, docset.RestoreMap);
 
             return new ContributionInfo(commitsByFile, updateTimeByCommit, UserProfileCache.Create(userProfileCachePath));
         }
 
+        private static string GetRestorePathFromUrl(string url, RestoreMap restoreMap)
```

Comment:
>GetRestorePathFromUrl [](start = 30, length = 21)

consider moving these logic to restore map, this layer only need a function that takes a path (URL or file) and returns the desired file path.

Reply:
restore map already has that method, is method is more focusing the busness logic like empty handling and error handling

---
In reply to: [202253232](https://github.com/dotnet/docfx/pull/3065#discussion_r202253232) [](ancestors = 202253232)

Reply:
The error handling and file.exists check will be duplicated to a lot of places when you implement config extends, xrefmap lookup, moniker definition lookup, etc.

---
In reply to: [202302511](https://github.com/dotnet/docfx/pull/3065#discussion_r202302511) [](ancestors = 202302511,202253232)